\documentclass{bioinfo}
\copyrightyear{2015} \pubyear{2015}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Manuscript Category}
\usepackage{tikz}
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{float}
\usepackage{url}
\usepackage{multirow}

\begin{document}
\firstpage{1}

\subtitle{Genome analysis}

\title[popSTR]{popSTR: population-scale detection of STR variants}
\author[Kristmund\'{o}ttir \textit{et~al.}]{Sn\ae d\'{i}s Kristmundsd\'{o}ttir\,$^{\text{\sfb 1,}*}$, Brynja D. Sigurp\'{a}lsd\'{o}ttir\,$^{\text{\sfb 1,2}}$, Birte Kehr\,$^{\text{\sfb 1}}$ and Bjarni V. Halld\'{o}rsson\,$^{\text{\sfb 1,2}}$}
\address{$^{\text{\sf 1}}$deCODE genetics/Amgen, Reykjav\'{i}k, 101, Iceland and \\
$^{\text{\sf 2}}$Institute of Biomedical and Neural Engineering, Reykjav\'{i}k University, Reykjav\'{i}k, 101,
Icleand.}

\corresp{$^\ast$To whom correspondence should be addressed.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{\textbf{Motivation:}Microsatellites, also known as short tandem repeats (STRs) are short DNA sequences containing repeated motifs ranging from 2-6 bases. 
There are about 700.000 microsatellites on the human reference genome covering almost 1\% of it \citep{eSTRgymrek} with the most common form being dinucleotide repeats.
The range of applications for microsatellite analysis is very wide and includes among other things medical genetics, forensics and genetic genealogy. However, microsatellite variations are rarely considered in whole-genome sequencing studies in large due to a lack of tools capable of analyzing them \citep{Duitama2014}.\\
\textbf{Results:} Here we present a microsatellite genotyper which is faster and more accurate than others previously presented. In order to accomplish this we utilize two things.
First, we reduce the amount of sequencing data necessary for creating microsatellite profiles when using previously aligned sequencing data. This is done by filtering the input to contain only reads aligned to known microsatellite locations and unaligned reads as these should be the ones useful for profiling.
Second, we use population information to train microsatellite and individual specific error profiles. 
Combining these two procedures we are able to give a practical implementation of microsatellite genotyping which is both faster and more accurate than previously presented solutions.\\
\textbf{Availability:} Code available on Github\\
\textbf{Contact:} \href{snaedis.kristmundsdottir@decode.is}\\
\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics}
online.}

\maketitle

\section{Introduction}
\label{s:intro}
Microsatellites (a.k.a. short tandem repeats, STRs) are short DNA sequences containing a repeated motif of length 2-6 base pairs.
Microsatellites are one of the most abundant type of variation in the human genome, after Single Nucleotide Polymorphisms (SNPs) and Indels.  
Microsatellites have a mutation rate estimated between
$1 \cdot 10^{-4}$ and $1 \cdot 10^{-3}$ mutations per locus per generation \cite{sun2012direct}, much higher than the mutation rate 
estimated for SNPs~\cite{KongSNPrate} of $1.2 \cdot 10^{-8}$.  Their high mutation rate made microsatellites particularly alluring 
for genotyping during the linkage era~\cite{LinkagePaper}.
Due to their high mutation rate, the alleles of a microsatellite to vary greatly between individuals~\cite{sun2012direct} and,  
appart from identical twins, no pair of individuals alive today have the same combination of alleles for all microsatellites \cite{Highnam2013}.
Using relatively few microsatellites, it is possible to create a unique genetic profile for every individual~\cite{NHGRI1999},
making microsatellites appealing for applications such as forensic analysis~\cite{forensicsMsat}.
Despite their abundance and the increasing availability of whole genome sequencing data, microsatellites are however often neglected 
in GWAS studies~\cite{Gudbjartsson2015}.

The high mutation rate can be attributed to the repetitive structure of microsatellites, which causes a secondary DNA conformation that 
makes replication slippage events more likely than in other locations of the genome~\cite{Mirkin2007}. 
Replication slippage occurs during replication when the copy strand being created and the original template strand get shifted in their relative 
positions, causing a part of the template to either be copied twice or not copied at all (cf. Figure ~\ref{fig:RepSlipp}) and resulting 
in either an increase or decrease in the number of motif repeats~\cite{Brown2002}.

\begin{figure}[ht]
\centering
 \includegraphics[width=0.5\linewidth]{figures/replicationSlippage.eps}
  \caption[Replication Slippage]{An extra repeat element added because of replication slippage \cite{Brown2002}.}
 \label{fig:RepSlipp}
\end{figure}

Replication slippage can occur within individual cells, as well as when the sample is being analyzed.
A slippage event that occurs during replication of a sex cell results in a germline mutation and may be passed on to an offspring, while slippage events 
within other cells of the body lead to somatic mutations.  Slippage events also frequently occur in PCR amplification, a preprocessing step often performed
prior to sequencing or during sequencing itself.  As a result, the sequence reads of an individul contain both reads from its germline variants and 
reads resulting from slippage events, complicating the genotyping of microsatellites.

The genotyping of microsatellites is further complicated by the fact that their high mutation rate can make it difficult to align microsatellite reads to the correct
location on the genome; most popular read-to-reference aligners trade-off between the tolerance of insertions/deletions and running time.
Yet another complication is the length of the microsatellite, as aligners generally require a unique match to the genome to seed their alignment, reads that are fully 
contained within a microsatellite can often not be placed within the genome.  Further, reads that do not fully encompass the microsatellite and only contain a portion 
of the microsatellite can only give a lower bound on the number of repeats~\cite{Gymrek2012}. 

A number of methods have been developed to genotype microsatellites~\cite{VNTRseek,RepeatSeq,lobSTR}.   We present {\em popSTR}, a method capable of studying 
microsatellite ({\em STR}) variation within all individuals of a {\em pop}ulation simultaneously.  
Our primary emphasis is on determining genotype likelihoods; the likelihood of a genotype given the sequencing reads at the given location.  
Microsatellite mutation rates have been shown to vary greatly between microsatellites as well as between individuals~\cite{sun2012direct}. Consequently, our model allows 
for an error model specific to each microsatellite and individual being studied.

Our results indicate that our method is both faster and more accurate than previously described methods.
%Need to expand on above, once we have results

%\enlargethispage{12pt}

%\section{Approach} Is it necessary to include this section?

\begin{methods}
\section{Methods}
popSTR requires three inputs; a reference genome, sequencing data of the set of individuals (population) being studied and a list of microsatellite locations (markers) on the reference genome.  
We assume that the sequencing data is Illumina whole genome sequencing data, mapped to the reference genome and stored in BAM-files, with the of each individual stored in a seperate bam file.
A set of microsatellite locations can be determined by running Tandem Repeat Finder (TRF)~\cite{trf} on the reference genome.  The output of popSTR are; for each marker the set of alleles occurring 
in at least one individual in the population and the genotype likelihoods of all alleles of the marker for each individual.

We start by describing our mathematical model for the likelihood of observing read data given a marker's genotype and then show how this is used in our implementation to find genotype likelihoods.

\enlargethispage{6pt}
\subsection{Mathematical models}
We assume we are given a set, $R$, of reads from a single individual and a single microsatellite.
There are multiple sources of error need to be accounted for in our model; 
the (replication) slippage rate is dependent on the marker being considered as well as the individual, additionally some of the reads 
being considered may be the results of sequencing or mapping errors.  
Slippage has two forms; full motif slippage and partial motif slippage.  A full motif slippage is when the length of the slippage is an integer multiple of the length of the repeat motif of the 
microsatellite, all other slippages are considered to be partial motif slippage.  Following~\cite{lobSTR} we model these two types of slipage events seperately.
We assume a Poisson distribution for full motif slippage events and a geometric distribution for partial motif slippage events.
Sequencing and mapping errors are accounted for using a logistic regression classifiers for each microsatellite seperately. 
The classifier assigns a probability for each read of being an error read, i.e. the result of a mapping or sequencing error.

\subsubsection{Slippage model}
\label{subsubsec:modelSlipp}
%To determine a genotype for individual $j$ at microsatellite $i$ we start by looking at all reads for the given individual aligned to the given microsatellite and determine the 
%possible genotypes they offer. We then compute the likelihood of the reads given each genotype and pick the genotype that maximizes this likelihood. The expression used to compute 
%this likelihood for a set of reads $R = \{r_1,\dots,r_n\}$ and a genotype $gt$ is 
We let $i$ represent the marker being considered and $j$ represent the individual.
We assume that the reads in $R$ are independent and we focus on the case when we would like to determine the likelihood of the genotype $gt$.  
For each read $r\in R$ we assume we let $m_i(r)$ represent the number times the repeat motif of $i$ is repeated in $r$.

\begin{equation}
L(R|gt) = \prod_{r\in R} L(r|gt)
\end{equation}

We now show how to compute $L(r|gt)$, adding terms for each source of error successively to our model.
We first consider the case when the only sources of error full motif slippage events and read misclassification events.
In this case, we assume a Poisson distribution for the slippage with $\lambda = S_{ij}$.  We let $p^t(r)$ denote the probability that read $r$ 
is not the result of a misclassification event.  We represent alleles with the number of times the repeat motif $m_i$ is repeated.  
Given an allele, $A$, we compute $x_r(A)$ the number of slippage events needed to explain $r$ with $A$ as $x_r(A) = A-m_i(r)$.
This gives the following expression for a homozygous genotype $(A,A)$.
\begin{equation}
L(r|A,A) = p^t(r) \cdot pois(x_r(A);S_{ij})
\label{eq:likelihood}
\end{equation}
For a heterozygous genotype $(A,B)$ we assume that each allele is drawn with equal probability:
\begin{multline}
L(r|A,B) = \\
p^t(r) \cdot 
(\frac{1}{2} \cdot pois(x_r(A);S_{ij})+ \frac{1}{2} \cdot pois(x_r(B);S_{ij}))
\end{multline}
%Perhaps skip the two above expressions
The above expression assigns a very small likelihood for reads that are not the results of slippage events.  With probability $1-p_s^r$ the read being considered is an error read, in this case we assume 
that each of the other reported alleles is equally likely. We let $n^i_{alleles}$ 
%Is the _{alleles} really necessary?
be the number of alleles present in the population for microsatellite $i$ and refine our expression for $L(r|gt)$ as follows:
\begin{multline}
L(r|A,B) = p^t(r) \cdot (\frac{1}{2} \cdot pois(x_r(A);S_{ij})+ \\
\frac{1}{2} \cdot pois(x_r(B);S_{ij})) + \frac{1-p^t(r)}{n^i_{alleles}}
\end{multline}
Slippage events are more likely to delete repeat units than insert.  To account for this, we further refine our model and add a parameter, $p_d$, representing the probability 
that if a slippage event occurs this event results in a deletion of a motif.  Given an allele $A$ and a read $r$ we compute $a_r^A$ as $p_d$ if $x_r(A) \le 0$ and $1-p_d$ if $x_r(A) > 0$.  
Our refined model then becomes:


\begin{multline}
L(r|A,B) = p^t(r) \cdot \\
(\frac{1}{2} \cdot pois(x_r(A);S_{ij}) \cdot a_r^A+ \frac{1}{2} \cdot pois(x_r(B);S_{ij})\cdot a_r^B)\\
+ \frac{1-p^t(r)}{n^i_{alleles}}
\label{eq:likelihood3.5}
\end{multline}

Finally we account for partial motif slippage events. Following the model presented in lobSTR~\cite{Gymrek2012} we model these using 
a geometrical distribution where $p = \frac{1}{\bar{x}+1}$ and $\bar{x}$ is the average step size modulo the length of the repeat motif.
%WTF is ``average step size modulo the length of the repeat motif''?
To reflect this we update our final model is:
%This formula needs to be cleaned
\begin{multline}
    L(r_s|A,B) = p(r^t_s) \cdot \\
    \biggl(\frac{1}{2} \cdot pois(floor(x_A);S_{ij}) \cdot geom(x_A-floor(x_A);p) \cdot a_1 \\ + \frac{1}{2} \cdot pois(floor(x_B);S_{ij})\cdot geom(x_A-floor(x_A);p) \cdot a_2 \biggr) \\
    + \frac{1-p^t(r)}{n^i_{alleles}}
    \label{eq:likelihood4}
\end{multline}
%This final expression combines the power obtained through the logistic regression classifier, the estimation of slippage rates at microsatellites 
%and individuals and includes both the Poisson component and th geometric distribution component of the slippage. By adding the probability that a 
%read is an error to the number it contributes to the likelihood we raise the tolerance for error reads. This makes it possible to call a homozygous 
%genotype despite having perhaps a number of reads supporting another allele if these other reads get poor results from the logistic regression classifier.

\subsubsection{Slippage rate estimation}
The frequency of slippage events varies between microsatellites. To account for this we estimate a marker specific slippage rate.   Assuming we know which reads are the results of a slippage 
event, we can estimate the slippage rate at microsatellite $i$ by dividing the number reads at the microsatellite marked as error reads by the total number of reads aligned to the microsatellite.
The following expression is used to estimate $S^M_i$, the slippage rate at microsatellite $i$
\begin{equation}
    S^{M}_i = \frac{\sum n^e_i}{\sum n_i}
\label{eq:SmInit}
\end{equation}
where $n^e_i$ represents the number of reads at microsatellite $i$ marked as being the result of a slippage event and $n_i$ represents the total number of reads assigned to microsatellite $i$.
This expression however ignores the fact that individuals may have different slippage rates.
We assume that the slippage of a marker $i$ at an individual $j$, $S_{ij}$ is a composite of a marker specific slippage rate, $S^M_i$ and an individual specific slippage rate $S^P_j$.
\begin{equation}
    S_{ij} = S^{M}_i+S^{P}_j 
\end{equation}
We denote the probability that a read assigned to microsatellite $i$, marked as an error read is actually reporting a true allele as $p^e_i$. 
%Error allele reporting a true allele?  
%How is p^e_i computed?
%Is p_i = p^t(r)? 
The probability that a read aligned to microsatellite $i$, regardless of how it is marked, reports a true allele is represented by $p_i$. For reads grouped by individuals we let $p_{ij}$ denote the probability of reporting a true allele for a read aligned to microsatellite $i$ from individual $j$. To estimate the slippage rate at microsatellite $i$ we then use the following expression
\begin{equation}
S^{M}_i = \frac{\sum p^e_i}{\sum p_i} - 
\sum_j S^{P}_j \cdot \frac{\sum p_{ij}}{\sum p_i}
\label{eq:sMjEq}
\end{equation}
In the expression used for estimating the slippage of individual $j$ we let $p^e_j$ denote the probability that an error read from individual $j$ reports a true allele and $p_j$ denote the probability that a read from individual $j$ reports a true allele, regardless of how it is marked. So the expression becomes
\begin{equation}
S^{P}_j = \frac{\sum p^e_j}{\sum p_j} - 
\sum_i S^{M}_i \cdot \frac{\sum p_{ij}}{p_j}
\label{eq:sPjEq}
\end{equation}
Although this estimate improves the previous expression it results in very high slippage rate estimates in some cases because all individuals contribute equally to the slippage at each position and vice versa without regard to their variance. If an individual or a microsatellite has a high variance they will then cause an overestimation of the slippage rate.
To decrease the weight of individuals and microsatellites with noisy reads in the slippage rate estimates we update the estimate again and weigh the contributions of each individual and microsatellite with the inverse of their variance. This gives microsatellites and individuals with low variance higher weights in the estimate and should thus make it more stable and reliable. 
The slippage rate at a microsatellite essentially represents the probability of getting an slippage error read at the given microsatellite and it is a binary distributed variable. We define the variance of a microsatellite as its slippage rate and use the definition of a variance for one trial on a binary distributed variable, given by
\begin{equation}
v = p(1-p)
\end{equation}
After this update, the expression for the slippage rate at microsatellite $i$ becomes 
\begin{equation}
S^{M}_i = \frac{\sum p^e_i}{\sum p_i} - 
\sum_j S^{P}_j \cdot 
\frac{ \frac{1}{S^{P}_j(1-S^{P}_j)} \cdot \frac{\sum p_{ij}}{\sum p_i}}{\sum_j \frac{1}{S^{P}_j(1-S^{P}_j)} \cdot \frac{\sum p_{ij}}{\sum p_i}}
\label{eq:sMiEq2}
\end{equation}
and similarly the slippage rate for individual $j$ becomes 
\begin{equation}
S^{P}_j = \frac{\sum p^e_j}{\sum p_j} - 
\sum_i S^{M}_i \cdot
\frac{ \frac{1}{S^{M}_i(1-S^{M}_i)} \cdot \frac{\sum p_{ij}}{\sum p_j}}
{\sum_i \frac{1}{S^{M}_i(1-S^{M}_i)} \cdot \frac{\sum p_{ij}}{\sum p_j}}
\label{eq:sPjEq2}
\end{equation}
Using these estimates we include the overall slippage rate of the individual or microsatellite being considered along with the contribution of each microsatellite to the individual and the contribution of each individual to the microsatellite. These contributions are however weighed with the inverse of their variance so noisy individuals or microsatellites will not skew the estimates.
\end{methods}

\section{Algorithm}
popSTR has 3 steps. During the first step we identify the reads useful for genotyping, compute their attributes and initialize genotypes. In the second step we compute an individual-specific slippage rate, quantifying the probability of stutter noise reads in this individual. In the third and last step we compute a marker specific slippage rate to quantify the probability of stutter noise at this marker, we train a logistic regression classifier for each marker and we perform genotyping.

\subsection{Read selection}
    The read selection simultaneously scans through the sequencing data and the microsatellite location file and compares all reads to the current microsatellite coordinates. If a read intersecting a microsatellite is found it is stored along with its mate. When we reach reads with an alignment location greater than the current microsatellite coordinates we move on to the next microsatellite.
    
    \noindent When a possibly useful read has been identified we check whether it contains the microsatellite it was aligned to. If it does, we align the flanking sequences of the repeat to the flanking sequences of the microsatellite, which we get from the reference genome and compute various attributes.
    After scanning through all reads we initialize genotypes based on frequencies.
\subsection{Logistic regression and genotyping}
    We train a logistic regression classifier for each microsatellite. The purpose is to account for errors not caused by stutter noise. We train the classifiers using only reads reporting true alleles and error reads not thought to result from stutter noise. Each read is assigned a probability of reporting a true allele by classifying it using the classifier trained for the microsatellite it covers. By including this probability in the slippage rate estimation expressions and in the expression used to determine the genotype and we attempt identify non-slippage error reads as well.
\section{Implementation}
PopSTR was written in C++ using the sequence analysis library SeqAn which allows for easy reading and manipulation of data stored in BAM-files \cite{Doering2008}.
\subsection{Read selection}
    We use the fact that the sequencing data has already been aligned and it is therefore possible to consider only reads aligned to known microsatellite locations along with unaligned reads. The unaligned reads are included since we don't know if they contain microsatellites or not. Sequences already aligned to non-microsatellite sequences are unlikely to be useful while sequences that are unaligned may in fact contain a microsatellite but have not been aligned because they are too different from the reference. We ran the Tandem Repeat Finder to determine the microsatellite locations \cite{TRF}.
    
    As mentioned before, when a useful read has been detected we align the flanking sequences of the repeat in the read to the microsatellite flanking sequences, which we get from the reference genome. The quality of this alignment is used as an indicator of how reliable the read is when performing genotyping. The purity of the repeat sequence in the read must also reach a minimum value for the read to be used and this minimum value is relative to the purity of the microsatellite sequence in the reference.
    
    To address the challenge presented by reads only able to give a lower bound on the number of repeats we allow for a user-specified maximum $m_l$. If the repeat occurs at either end of a read or spans the entire read and the base-pair length of the repeat is larger than $m_l$ we report the number of repeats as:
    \begin{equation}
        \frac{m_l}{lengt(repeatMotif)}
    \end{equation}
    These cases however require higher purity values from the the repeat and a minimum alignment score from the flanking sequences when they are available.(see exact numbers in Table ~\ref{table:numConditions}) 
    This gives the user the option to lump all alleles above a certain length into a "greater-than" allele.
    
    The following table summarizes the numerical values used as conditions in various cases during the read selection.
    \begin{center}
        \begin{table}[H]
            \caption{Minimum numeric values when identifying useful reads}
            \begin{tabular}{p{0.5\linewidth}|p{0.5\linewidth}}
                \hline
                \textbf{Name - Condition} & \textbf{Minimum value} \\ \hline
                Purity - standard case & 0.75*(ref. purity) \\ \hline
                Purity - no flanking & 0.85*(ref. purity) \\ \hline
                Purity - one sided flanking & 0.8*(ref. purity) \\ \hline
                Matching flank-bases fraction-one sided flanking & 0.7 \\ \hline
                Number of repeats - motiflength = 2 & 4 \\ \hline
                Number of repeats - motiflength = 3 & 3 \\ \hline
                Number of repeats - motiflength $\in {4,5,6}$ &  2\\ 
                \hline
            \end{tabular}
            \label{table:numConditions}
        \end{table}
    \end{center}
    
    We throw away low quality reads, i.e. the ones that fail platform or vendor quality checks and reads that are PCR or optical duplicates. This should increase the result quality since the reads being removed could possibly decrease the genotyping accuracy.
    When we have scanned all of the sequencing data and computed attributes for all useful reads at every microsatellite, we initialize genotypes based on frequencies and write both the attributes and the genotype initialization to an output file.
\subsection{Slippage computation - Individual and marker specific}
    The slippage for individual $j$ is initialized by dividing the number of reads reporting an allele not in the initialized genotype by the total number of reads.
    \begin{equation}
            S^{P}_j = \frac{\sum n^e_j}{\sum n_j}
    \end{equation}
    When marker specific slippage rates have been computed and probabilities from logistic regression assigned to all reads, we update our estimate according to equation ~\ref{eq:sPjEq2}
    All probabilities in the marker slippage estimate expression given in equation ~\ref{eq:sMiEq2} are initialized as 0.95 but in later iterations they are computed for each read by plugging its attributes into the logistic regression classifier of the microsatellite they were aligned to.
\subsection{Logistic regression classifier and genotyping}
    The logistic regression attributes used are various but all have in common their ability to indicate the quality of a read in one way or the other. These are summarized in Table ~\ref{table:attributes}. 

    \begin{table*}[ht]
        \caption{The attributes used as control variables in the Logistic regression classification.}
        \begin{tabular}{p{0.5\linewidth}|p{0.5\linewidth}}
            \hline
            \textbf{Attribute} & \textbf{Definition} \\ \hline
            Quality score & Map quality score of the aligned read. \\ \hline
            Purity & \# of repeats in the microsatellite sequence/ 
            \# of expected repeats in the microsatellite sequence \\ \hline
            Number over 20 in & The number of bases in the read with a PHRED quality over 20 in the microsatellite sequence. \\ \hline
            Number over 20 after & The number of bases in the read with a PHRED quality over 20 coming after the microsatellite sequence. \\ \hline
            Edit distance of mate & Edit distance to the reference, including ambiguous bases but excluding clipping. \\ \hline
            Alignment score before & Alignment score of sequence before the microsatellite to the reference. \\ \hline
            Alignment score after & Alignment score of sequence after the microsatellite to the reference. \\ \hline
            Was unaligned & Boolean value indicating if the read was unaligned. \\ \hline
            Location shift & Measures alignment shifting during the realignment of flanking sequences. \\ \hline
            Sequence length & Total length of the read-sequence. \\
            \hline
        \end{tabular}
        \label{table:attributes}
    \end{table*}
    
    Having assigned a probability to all reads and estimated the slippage rates at all microsatellites and for all individuals we now have what we need to determine the genotypes. To do this for individual $j$ at microsatellite $i$ we follow the process described in Section ~\ref{subsubsec:modelSlipp}. Probabilities of deletion or addition in Equation ~\ref{eq:likelihood4} were set as $p_d = 0.85$ and $p_a = 0.15$. 
\subsection{Data set}
 While training we iterate between steps II and III until convergence has been reached. In 2015, popSTR was run on around 8000 individuals for around 700.000 microsatellites. We chose a training set of 8303 microsatellites based on their good results from imputation into the Icelandic population. Training was performed using 703 individuals with high quality sequencing data.
 Using results from the training to directly estimate slippage rates for new individuals on these 8303 microsatellites allows us to skip the training process in later genotyping. 
 Comparison with lobSTR was done by choosing 10 random individuals from the 15.220 sequenced individuals available at the time. We chose 146 markers where inhouse capillary electrophoresis benchmark genotypes were available and compared our and lobSTR genotypes to those.
 To further investigate the quality of our genotypes we matched our microsatellites coordinates to output coordinates and indel alleles from the GATK genotype caller and compared the imputation results into the Icelandic populations for markers where matching was possible. 
 
\section{Results}
\subsection{Training}
During the training 5 iterations were performed until convergence was reached. Figures ~\ref{fig:pnSlipp} and ~\ref{fig:markerSlipp} show the distribution of slippage rate values for the training set.
\begin{figure}[H]
\centering
 \includegraphics[width=\linewidth]{figures/pnSlippageHist.png}
  \caption[Histogram of pn-slippage]{Histogram of slippage rates for the 703 individuals in the training set.}
 \label{fig:pnSlipp}
\end{figure}

\begin{figure}[H]
\centering
 \includegraphics[width=\linewidth]{figures/markerSlippageHist.png}
  \caption[Histogram of marker-slippage]{Histogram of slippage rates for the 8303 microsatellites in the training set.}
 \label{fig:markerSlipp}
\end{figure}

\subsection{Running popSTR on 15.220 individuals}
A total of 380.261 microsatellites were imputed into the Icelandic population.
First, we performed step I and identified useful reads and computed their attributes for all individuals at all markers. Next, we directly estimated individual slippage rates using only reads aligned to the 8303 microsatellites from the training set. We used the logistic regression classifiers from the training to assign probabilities to the reads and plugged the probabilities into equation ~\ref{eq:sPjEq2} along with the marker slippage rates from training. 
As a sanity check of the individual slippage rates we checked if they correlate significantly with age. The slippage rate was expected to increase with age due to somatic mutations that accumulate throughout a persons life. Due to a sequencing method batch effect observed in the data we considered only individuals sequenced using HiSeqX machines. Despite this we observed a batch effect as a result of moving from methods using PCR amplification to PCR-free methods. We therefore included a boolean for this property in the age correlation test. The results from that were very clear that slippage rate increases with age (p-value=8.21e-13)
Last we genotyped all individuals using Equation ~\ref{eq:likelihood4}.

\subsection{Comparison to lobSTR}
We compared the popSTR genotypes and the lobSTR genotypes to inhouse benchmark data obtained through capillary electrophoresis. The comparison was done on a set of 10 individuals chosen at random and 146 microsatellites chosen due to the availability of data. 
Comparison was performed both considering all genotypes as well as considering only genotypes where 10 or more reads were available. Table ~\ref{table:lobSTRcomp} summarizes the results. 
\begin{table}[H]
    \caption{Accuracy comparison of lobSTR and popSTR with and without coverage filtering.}
    \begin{tabular}{ c | c | c | c | c |}
         & \multicolumn{2}{c}{No coverage filter} & \multicolumn{2}{c}{coverage $ \geq 10$} \\
         \hline
         & \# genotypes & accuracy & \# genotypes & accuracy \\
         \hline
        popSTR & 1138 & 92.5\% & 969 & 95.1\% \\
        \hline
        lobSTR & 1135 & 86.8\% & 647 & 91.2\% \\
        \hline
    \end{tabular}
    \label{table:lobSTRcomp}
\end{table}
\subsection{Comparison to GATK}
Compare imputation info of my genotypes from the big run to GATK calls for microsatellites matched by Brynja's program.
Histogram like in Brynja's report on where my info is higher and where lobSTR info is higher

\section{Discussion}

Here I will discuss problems, improvements and results. Such as not identifying fewer than x repeats (depending on motif length) and trying to find the optimal value for $m_l$. Mention that association will be run on these results ? 









%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%






\section{Conclusion}
Here we have shown that when creating a microsatellite profile for an individual using previously aligned data it is possible to significantly decrease the running time by considering only reads that are either aligned to a known microsatellite location or not aligned at all. This type of filtering dismisses a large portion of the data immediately without significantly effecting the microsatellite profile. To do this we used the alignment locations of reads along with attributes indicating undesirable qualities.


\section*{Acknowledgements}
deCODE, Amgen ??? 

%\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
\bibliographystyle{plain}
\bibliography{document}
\end{document}
